apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,322 @@
+        /// </summary>
+        public string JsonCredentials { get; private set; }
+
+        /// <summary>
+        /// The token access method to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public Func<string, CancellationToken, Task<string>> TokenAccessMethod { get; private set; }
+
+        /// <summary>
+        /// The call invoker to use, or null to create the call invoker when the client is built.
+        /// </summary>
+        public CallInvoker CallInvoker { get; private set; }
+
+        /// <summary>
+        /// Creates a new instance with no settings.
+        /// </summary>
+        protected ClientBuilder()
+        {
+        }
+        
+        /// <summary>
+        /// Creates a clone of the given builder, with respect to settings in this class.
+        /// </summary>
+        /// <param name="builder">The builder to clone</param>
+        protected ClientBuilder(TBuilder builder)
+        {
+            Endpoint = builder.Endpoint;
+            ChannelCredentials = builder.ChannelCredentials;
+            Scopes = builder.Scopes;
+            CredentialsFile = builder.CredentialsFile;
+            JsonCredentials = builder.JsonCredentials;
+            CallInvoker = CallInvoker;
+            ChannelCredentials = null;
+        }
+
+        /// <summary>
+        /// This method should be called by all "With" methods that change the credentials.
+        /// If a subclass supports additional forms of credentials, it should override this method
+        /// to clear them and call this implementation.
+        /// </summary>
+        protected virtual void ClearCredentials()
+        {
+            CredentialsFile = null;
+            JsonCredentials = null;
+            TokenAccessMethod = null;
+            CallInvoker = null;
+            ChannelCredentials = null;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified endpoint.
+        /// </summary>
+        /// <param name="endpoint">The endpoint to connect to, or null to use the default endpoint.</param>
+        /// <returns>A builder with the specified endpoint.</returns>
+        public TBuilder WithEndpoint(ServiceEndpoint endpoint)
+        {
+            var clone = Clone();
+            clone.Endpoint = endpoint;
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified credential scopes.
+        /// </summary>
+        /// <param name="scopes">The scopes to use when creating credentials, or null to use the default set of scopes.</param>
+        /// <returns>A builder with the specified scopes.</returns>
+        public TBuilder WithScopes(IEnumerable<string> scopes)
+        {
+            var clone = Clone();
+            clone.Scopes = scopes?.ToList().AsReadOnly();
```

Comment:
nit: Is `.AsReadOnly()` necessary? It's only ever exposed as an `IReadOnlyList<>`

Reply:
It could be cast.